lib/core/utils/selector-cache.js

```diff
@@ -0,0 +1,196 @@
+import { matchesExpression } from './matches';
+import tokenList from './token-list';
+
+// since attribute names can't contain whitespace, this will be
+// a reserved list for ids so we can perform virtual id lookups
+const idsKey = ' [idsMap]';
+let selectorMap = {};
+
+/**
+ * Non-tag selectors use `*` for the tag name so a global selector won't have any other properties of the expression.
+ * @param {Object} exp Selector Expression
+ * @returns {Boolean}
+ */
+function isGlobalSelector(exp) {
+  return exp.tag === '*' && !exp.attributes && !exp.id && !exp.classes;
+}
+
+/**
+ * Save a selector and vNode to the selectorMap
+ * @param {String} key
+ * @param {VirtualNode} vNode
+ * @param {Object} map
+ */
+function cacheSelector(key, vNode, map = selectorMap) {
+  map[key] = map[key] || [];
+  map[key].push(vNode);
+}
+
+/**
+ * Inner join two arrays (find all nodes in a that are also in b).
+ * @param {Mixed[]} a
+ * @param {Mixed[]} b
+ * @returns {Mixed[]}
+ */
+function innerJoin(a, b) {
+  if (!b.length) {
+    return a;
+  }
+
+  return a.filter(node => b.includes(node));
+}
+
+/**
+ * Cache selector information about a VirtalNode.
+ * @param {VirtualNode} vNode
+ */
+export function cacheNodeSelectors(vNode) {
+  if (vNode.props.nodeType !== 1) {
+    return;
+  }
+
+  // cache the selector map to the root node of the tree
+  if (vNode.nodeIndex === 0) {
+    selectorMap = {};
+    vNode._selectorMap = selectorMap;
+  }
+
+  cacheSelector(vNode.props.nodeName, vNode);
+  cacheSelector('*', vNode);
+
+  vNode.attrNames.forEach(attrName => {
+    // element ids are the only values we'll match
+    if (attrName === 'id') {
+      selectorMap[idsKey] = selectorMap[idsKey] || {};
+      tokenList(vNode.attr(attrName)).forEach(value => {
+        cacheSelector(value, vNode, selectorMap[idsKey]);
+      });
+    }
+
+    cacheSelector(`[${attrName}]`, vNode);
+  });
+}
+
+/**
+ * Get nodes from the selector cache that match the selector.
+ * @param {VirtualTree[]} domTree flattened tree collection to search
+ * @param {Object} expressions
+ * @param {Function} filter function (optional)
+ * @return {Mixed} Array of nodes that match the selector or undefined if the selector map is not setup
+ */
+export function getNodesMatchingExpression(domTree, expressions, filter) {
+  // check to see if the domTree is the root and has the selector
+  // map. if not we just return and let our QSA code do the finding
+  const selectorMap = domTree[0]._selectorMap;
```

Comment:
It's a little confusing you're reusing this variable name. You now have this name both in the function scope and in the file scope.